src/Features/LanguageServer/Microsoft.CodeAnalysis.LanguageServer/HostWorkspace/HostDiagnosticAnalyzerProvider.cs

```diff
@@ -3,15 +3,72 @@
 // See the LICENSE file in the project root for more information.
 
 using System.Collections.Immutable;
+using System.Reflection;
 using Microsoft.CodeAnalysis.Diagnostics;
 using Microsoft.CodeAnalysis.Workspaces.ProjectSystem;
+using Roslyn.Utilities;
 
 namespace Microsoft.CodeAnalysis.LanguageServer.HostWorkspace;
 internal class HostDiagnosticAnalyzerProvider : IHostDiagnosticAnalyzerProvider
 {
+    /// <summary>
+    /// The directory where the analyzers are located
+    /// </summary>
+    public string? AnalyzerDirectory { get; set; }
+
+    private const string RazorVsixExtensionId = "Microsoft.VisualStudio.RazorExtension";
+    private static readonly HashSet<string> s_razorSourceGeneratorAssemblyNames = new[] {
+        "Microsoft.NET.Sdk.Razor.SourceGenerators",
+        "Microsoft.CodeAnalysis.Razor.Compiler.SourceGenerators",
+        "Microsoft.CodeAnalysis.Razor.Compiler",
+    }.ToHashSet<string>();
+
+    public HostDiagnosticAnalyzerProvider(string? analyzerDirectory)
+    {
+        AnalyzerDirectory = analyzerDirectory;
+    }
+
     public ImmutableArray<(AnalyzerFileReference reference, string extensionId)> GetAnalyzerReferencesInExtensions()
     {
-        // Right now we don't expose any way for the extensions in VS Code to provide analyzer references.
-        return ImmutableArray<(AnalyzerFileReference reference, string extensionId)>.Empty;
+        if (AnalyzerDirectory == null)
+        {
+            return ImmutableArray<(AnalyzerFileReference reference, string extensionId)>.Empty;
+        }
+
+        var analyzerReferences = new List<(AnalyzerFileReference, string)>();
+
+        // Get all the .dll files in the directory
+        var analyzerFiles = Directory.GetFiles(AnalyzerDirectory, "*.dll");
+
+        foreach (var analyzerFile in analyzerFiles)
+        {
+            // Create an AnalyzerFileReference for each file
+            var analyzerReference = new AnalyzerFileReference(analyzerFile, new SimpleAnalyzerAssemblyLoader());
+
+            if (s_razorSourceGeneratorAssemblyNames.Any(
+                name => analyzerReference.FullPath.Contains(name, StringComparison.OrdinalIgnoreCase)))
```

Comment:
Can this be `s_razorSourceGeneratorAssemblyNames.Contains(Path.GetFilenameWithoutExtension(analyzerFile))`?

Seems like that would be more efficient, and would allow moving the `analyzerReference` creation to inside the `if` statement, and avoid creating lots of new objects unnecessarily. Or does the `FullPath` property do some other logic?

Reply:
I simplified the logic. No need to check against the list, instead passing down directly the full path to the razor source generator.